### PR TITLE
Issue #2212: Fix issues in Bookkeeper Docker image that prevented containers from starting up

### PR DESCRIPTION
**Pull request by ravisharda**
Descriptions of the changes in this PR:

### Motivation

This PR modifies the Apache Bookkeeper Docker image to fix issues that were causing errors upon container bootstrap. The containers would exit soon after they were launched. See issue #2212 for a description of such an error. 

Note that the problems that are fixed in this PR were observed:

* when launching containers using both Docker Compose and Kubernetes.  
* when we were trying to upgrade the image to: `4.9.2`. It is highly likely that the issue is observed in other versions (except for `4.7.3`) too.  
* when launching both a standalone container as well as a cluster of three containers.

### Changes
The major changes made in this PR are as follows: 
* Updates the `Dockerfile` to install `zk-shell`. 
* Update the `init_bookie.sh` file to:
   * Use `zk-shell` instead of `/opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain` command that doesn't work. 
  * Use `opt/bookkeeper/bin/bookkeeper shell initnewcluster` for initializing the cluster instead of the previously used command that did not work. 
  * Increase the time a container waits for an in-flight `initnewcluster` operation. 
  * Make the comments more descriptive.
* Modifies `bin/common.sh` to handle the condition when file `/proc/sys/net/ipv6/bindv6only` is missing in the system. This can prevent a container from starting up in some cases. We have seen this issue on some Kubernetes-based environments. 
* Fixes errors in `docker-compose.yml` file. 

*Note:* Some of the changes made in this PR are modeled after changes made by @sijie for `v4.7.2` in PR #1666 .

### Master Issue
#2212 

> ---
> In order to uphold a high standard for quality for code contributions, Apache BookKeeper runs various precommit
> checks for pull requests. A pull request can only be merged when it passes precommit checks. However running all
> the precommit checks can take a long time, some trivial changes don't need to run all the precommit checks. You
> can check following list to skip the tests that don't need to run for your pull request. Leave them unchecked if
> you are not sure, committers will help you:
>
> - [ ] [skip bookkeeper-server bookie tests]: skip testing `org.apache.bookkeeper.bookie` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server client tests]: skip testing `org.apache.bookkeeper.client` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server replication tests]: skip testing `org.apache.bookkeeper.replication` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server tls tests]: skip testing `org.apache.bookkeeper.tls` in bookkeeper-server module.
> - [ ] [skip bookkeeper-server remaining tests]: skip testing all other tests in bookkeeper-server module.
> - [ ] [skip integration tests]: skip docker based integration tests. if you make java code changes, you shouldn't skip integration tests.
> - [ ] [skip build java8]: skip build on java8. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> - [ ] [skip build java11]: skip build on java11. *ONLY* skip this when *ONLY* changing files under documentation under `site`.
> ---

> ---
> Be sure to do all of the following to help us incorporate your contribution
> quickly and easily:
>
> If this PR is a BookKeeper Proposal (BP):
>
> - [ ] Make sure the PR title is formatted like:
>     `<BP-#>: Description of bookkeeper proposal`
>     `e.g. BP-1: 64 bits ledger is support`
> - [ ] Attach the master issue link in the description of this PR.
> - [ ] Attach the google doc link if the BP is written in Google Doc.
>
> Otherwise:
> 
> - [ ] Make sure the PR title is formatted like:
>     `<Issue #>: Description of pull request`
>     `e.g. Issue 123: Description ...`
> - [ ] Make sure tests pass via `mvn clean apache-rat:check install spotbugs:check`.
> - [ ] Replace `<Issue #>` in the title with the actual Issue number.
> 
> ---

### CODE REVIEW

**Comment by eolivelli** [comment]
@dmercuriali PTAL

**Comment by nicknezis** [comment]
I think this might fix issues I've had when trying to run Bookkeeper in Kubernetes.

**Comment by nicknezis** [comment]
Ran a local test with Heron Helm chart in Minikube using a locally built copy of this branch. My Bookkeeper startup was much more consistent and stable. Thank you.

**Comment by eolivelli** [comment]
run integration tests
run bookkeeper-server replication tests

**Comment by eolivelli** [comment]
@ravisharda It looks to me that the patch is in good shape
is this patch ready for final review ?
I have forced a new run of Travis and other CI tests.

**Comment by eolivelli** [comment]
@jiazhai @sijie @caiok PTAL

**Comment by eolivelli** [comment]
Travis is failing for:
org.apache.httpcomponents-httpclient-4.4.1.jar unaccounted for in LICENSE
org.apache.httpcomponents-httpcore-4.4.1.jar unaccounted for in LICENSE
org.apache.thrift-libthrift-0.9.3.jar unaccounted for in LICENSE

this is weird, because this patch is not changing the list of jars that we are bundling with the binary distribution

**Comment by ravisharda** [comment]
> @ravisharda It looks to me that the patch is in good shape
> is this patch ready for final review ?
> I have forced a new run of Travis and other CI tests.

Yes, I think it is ready for review. There is an issue that Pedro had reported in Slack that needs to be looked at, but IMO that shouldn't hold up the review. 

> Travis is failing for:
> org.apache.httpcomponents-httpclient-4.4.1.jar unaccounted for in LICENSE
> org.apache.httpcomponents-httpcore-4.4.1.jar unaccounted for in LICENSE
> org.apache.thrift-libthrift-0.9.3.jar unaccounted for in LICENSE
> 
> this is weird, because this patch is not changing the list of jars that we are bundling with the binary distribution

Yes, not sure why this is happening. Could these jars be coming as transitive dependencies via `zk-shell`?

**Comment by eolivelli** [comment]
@ravisharda Integration tests are still failing with this error:
```
2019-12-22 18:23:04,358 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,359 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,359 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,360 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,360 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,361 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,361 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,362 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,363 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,363 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,364 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,365 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
2019-12-22 18:23:04,366 - INFO  - [testcontainers-netty-1-21:Slf4jLogConsumer@32] - STDERR: /opt/bookkeeper/scripts/init_bookie.sh: line 26: zk-shell: command not found
``


check
https://builds.apache.org/job/bookkeeper_precommit_integrationtests/2506/testReport/junit/org.apache.bookkeeper.tests.integration.cluster/SimpleClusterTest/test001_StartBookie_on_test001_StartBookie_org_apache_bookkeeper_tests_integration_cluster_SimpleClusterTest_/

**Comment by eolivelli** [comment]
@ravisharda don't worry.
The problem is not on your patch, I have sent a dummy PR to demostrate the issue

https://github.com/apache/bookkeeper/pull/2222

I have sent an email to dev@bookkeeper.apache.org mailing list.
We will fix the problem soon.

Anyway we need some other committer to validate the patch

cc @fpj

**Comment by eolivelli** [comment]
@ravisharda we will be able to commit this patch once https://github.com/apache/bookkeeper/pull/2228 lands to master and fixes Travis build.

I am sorry for this taking so long.

@jiazhai  can you please take a look to the patch, as you were the main contributor of Dockerfile ?

**Comment by eolivelli** [comment]
retest this please

**Comment by eolivelli** [comment]
rebuild java11
run bookkeeper-server client tests

**Comment by eolivelli** [comment]
@ravisharda can you please rebase to latest master?
It seems that Travis did not pick up #2228

**Comment by eolivelli** [comment]
I have merged current master, in order to see a green build :-)

**Comment by eolivelli** [comment]
retest this please

**Comment by eolivelli** [comment]
@ravisharda it seems that Jenkins doesn't want to test this PR anymore :-( 

I am closing and reopening the PR in order to trigger Jenkins again

**Comment by eolivelli** [comment]
@ravisharda I am sorry.
Now Travis is passing but Jenkins isn't triggered anymore.
Can you please close this PR and create a new one?
Maybe squashing all of the commits and rebasing to latest master?

I hope we can land this patch soon, it will unblock a few other pendings works

**Comment by eolivelli** [comment]
It seems that Jenkins started building this patch.
I hope the tasks will complete with success

**Comment by eolivelli** [comment]
rebuild java11

**Comment by ravisharda** [comment]
Thanks, @eolivelli! Do you still want me to close this PR and create a new one?

**Comment by eolivelli** [comment]
No.
Jenkins is working now.
I will kick the integration tests again.

**Comment by eolivelli** [comment]
run integration tests

**Comment by ravisharda** [comment]
@eolivelli I had a quick look at the test failures. It is strange that `zk-shell` is missing in the integration test, even though it is getting installed by this line: 
https://github.com/apache/bookkeeper/blob/a7855d994ee70b22d12223dc14da602da9b26196/docker/Dockerfile#L52

Besides, when creating a Bookkeeper cluster manually using Docker Compose, the containers did come up properly. 

Could it be that the environment where integration tests run is not able to run the following lines that install `ZK-Shell`? If that is not the case, I might need to dig deeper into this. 

https://github.com/apache/bookkeeper/blob/a7855d994ee70b22d12223dc14da602da9b26196/docker/Dockerfile#L50-L53

cc / @sijie

**Comment by eolivelli** [comment]
@ravisharda yes it sounds really weird.

but it looks like that not all of the tests failed:
https://builds.apache.org/job/bookkeeper_precommit_integrationtests/2525/testReport/

only the ones about the "Stream Storage Service".

You can run them with 
`mvn clean install -Dstream -Pdocker`

**Comment by eolivelli** [comment]
Maybe it is a problem in Arquillian or TestContainers.
It needs debug.

**Comment by eolivelli** [comment]
This is actually the command:
```
cd tests/cluster
mvn test -Dstream -DintegrationTests -Dtest=SimpleClusterTest

I am trying to update TestContainers, but without success
```

**Comment by ravisharda** [comment]
Thanks, @eolivelli. I was able to reproduce the issue (using the following steps), for one of the integration tests that have failed. This will need a bit of debugging. I was not able to spend any cycles on this today but will try to get to it as soon as I can. 

```
$ mvn clean install -DskipTests -Pdocker
$ cd tests/integration/cluster
$ mvn test -Dstream -DintegrationTests -Dtest=SimpleClusterTest#test001_StartBookie
```

**Comment by eolivelli** [comment]
@ravisharda do you have news ?

**Comment by ravisharda** [comment]
> @ravisharda do you have news ?

@eolivelli I haven't gotten a chance to get back to it yet.

**Comment by eolivelli** [comment]
@ravisharda do you have time to complete this PR ?
If your fix works on Pravega codebase then we have good chance to see it working here. 

We have fixed a few problems on integration tests.
Maybe you just only have to rebase to current master

**Comment by ravisharda** [comment]
Sure @eolivelli, I just rebased the branch and pushed the changes.

**Comment by eolivelli** [comment]
This we have this problems:

[ERROR]   TableClientTest » ContainerLaunch Container startup failed
[ERROR] Tests run: 21, Failures: 5, Errors: 0, Skipped: 14
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:2.21.0:test (default-test) on project cluster: There are test failures.
[ERROR]

**Comment by ravisharda** [comment]
@eolivelli sorry, I am not getting any time to debug/resolve the test issues, which is keeping this PR open. Can someone help?

**Comment by eolivelli** [comment]
@atris @caiok @dmercuriali @rdhabalia do you have time to take a look ?

This change is mostly about porting Pravega.io enhancements to Community
and this patch is also unblocking the update to ZK 3.5.

BookKeeper does not work on JDK14 because ZK 3.4 does not work with JDK14

cc @sijie @merlimat  @fpj

**Comment by eolivelli** [comment]
Closed/Reopened in order to trigger tests again

File: docker/scripts/init_bookie.sh [hunk 1]
```diff
@@ -19,65 +19,57 @@
 # * See the License for the specific language governing permissions and
 # * limitations under the License.
 # */
-
 source ${SCRIPTS_DIR}/common.sh
 
 function wait_for_zookeeper() {
     echo "wait for zookeeper"
-    until /opt/bookkeeper/bin/bookkeeper org.apache.zookeeper.ZooKeeperMain -server ${BK_zkServers} ls /; do sleep 5; done
+    until zk-shell --run-once "ls /" ${BK_zkServers}; do sleep 5; done
```
Comment (eolivelli) [inline docker/scripts/init_bookie.sh:26]: In order to fix the integration tests I guess we should use the full path of zk-shell.
In fact the previous line used the full path

@dmercuriali can you please help fixing this patch fro @ravisharda  ?

**Comment by eolivelli** [comment]
I have created this branch with a possible fix over this branch
PTAL @ravisharda 

@dmercuriali once we merge this patch you will be able to complete the ZK upgrade as we will have zk-shell now

**Comment by eolivelli** [comment]
@ravisharda in #2320 all tests passed.
The patch should be good to go

**Comment by ravisharda** [comment]
> @ravisharda in #2320 all tests passed.
> The patch should be good to go

Thanks @eolivelli! I have merged the changes made in #2320 into this branch.

**Comment by eolivelli** [comment]
Committed to 4.11, 4.10 add 4.9 branches

thank you @ravisharda

**Pull request merged**